### PR TITLE
Mark Failed Message Send to Slave as Error & Retry

### PR DESCRIPTION
**Pull request by vincentbavitz**
### Fix 1
Throw error on message fail and automatically retry only for failed devices

### Fix 2
Leaving from Primary and Secondary quickly no longeer causes infinite sync loop

### Fix 3
Remove all user's devices from group as group admin on ``doUpdateGroup`` to avoid sync issues

### Fix 4
Removed archiving - oboslete

### CODE REVIEW

File: libtextsecure/outgoing_message.js [hunk 1]
```diff
@@ -508,16 +511,10 @@ OutgoingMessage.prototype = {
         this.errors.push(e);
       }
     });
-    await Promise.all(promises);
-    // TODO: the retrySend should only send to the devices
-    // for which the transmission failed.
```
Comment (neuroscr) [inline libtextsecure/outgoing_message.js:513]: maybe we should keep this comment?
```
// TODO: the retrySend should only send to the devices
// for which the transmission failed
```

**Comment by neuroscr** [comment]
why don't want want the automatic execution of promise array?

File: js/views/create_group_dialog_view.js [hunk 2]
```diff
@@ -179,18 +179,39 @@
...
+        return devices;
+      };
+
+      // Get all devices for notPresentInNew
+      const allDevicesOfMembersToRemove = await getDevicesForRemoved();
+
+      // If any extra devices of removed exist in newMembers, ensure that you filter them
+      const filteredMemberes = allMembers.filter(
+        member => !window.Lodash.includes(allDevicesOfMembersToRemove, member)
+      );
```
Comment (Bilb) [inline js/views/create_group_dialog_view.js:212]: don't get that part

File: js/views/create_group_dialog_view.js [hunk 3]
```diff
@@ -179,18 +179,39 @@
...
+
+        return devices;
+      };
+
+      // Get all devices for notPresentInNew
+      const allDevicesOfMembersToRemove = await getDevicesForRemoved();
+
+      // If any extra devices of removed exist in newMembers, ensure that you filter them
+      const filteredMemberes = allMembers.filter(
+        member => !window.Lodash.includes(allDevicesOfMembersToRemove, member)
```
Comment (Bilb) [inline js/views/create_group_dialog_view.js:211]: window.Lodash is also shortened as `_.` isn't it?

File: js/views/create_group_dialog_view.js [hunk 4]
```diff
@@ -179,18 +179,39 @@
...
       );
-      // would be easer with _.xor but for some reason we do not have it
+
+      // Filter out all linked devices for cases in which one device
+      // exists in group, but hasn't yet synced with its other devices.
+      const getDevicesForRemoved = async () => {
+        const promises = notPresentInNew.map(member =>
+          libloki.storage.getPairedDevicesFor(member)
+        );
+        const devices = window.Lodash.flatten(await Promise.all(promises));
```
Comment (Bilb) [inline js/views/create_group_dialog_view.js:201]: window.Lodash is also shortened as _. isn't it?

File: js/views/create_group_dialog_view.js [hunk 5]
```diff
@@ -179,18 +179,39 @@
...
       );
+
       const notPresentInNew = this.existingMembers.filter(
         m => !allMembers.includes(m)
       );
-      // would be easer with _.xor but for some reason we do not have it
+
+      // Filter out all linked devices for cases in which one device
+      // exists in group, but hasn't yet synced with its other devices.
+      const getDevicesForRemoved = async () => {
```
Comment (neuroscr) [inline js/views/create_group_dialog_view.js:197]: Why is this a function? Are we planning to use it more than once?

File: js/views/create_group_dialog_view.js (hunk 4 above)
Comment (vincentbavitz) [reply js/views/create_group_dialog_view.js:201]: Undefined in this block scope.
I can pull it out

File: js/background.js [hunk 6]
```diff
@@ -641,6 +641,53 @@
...
+          
+        // });
+        // const pairedDevices = 5;
+
+        // console.log('[vince] this.members:', this.get('members'));
+        // console.log('[vince] providedGroupUpdate:', providedGroupUpdate);
+        // console.log('[vince] groupUpdate:', groupUpdate);
+        
+        
+        console.log('[vince] doUpdateGroup: members:', members);
```
[the remark this answers is not in the document]
Comment (vincentbavitz) [reply js/background.js]: Was fixed several commits before comment

File: js/views/create_group_dialog_view.js (hunk 2 above)
Comment (vincentbavitz) [reply js/views/create_group_dialog_view.js:212]: Forcibly removes slave devices from member list if it's a secondary device AND its primary is in the chat. Solves various quirky bugs

File: js/views/create_group_dialog_view.js (hunk 5 above)
Comment (vincentbavitz) [reply js/views/create_group_dialog_view.js:197]: Allows us to cleanly await all in ``libloki.storage.getPairedDevicesFor(member)`` before ``window.doUpdateGroup`` is executed.

File: js/views/create_group_dialog_view.js (hunk 4 above)
Comment (vincentbavitz) [reply js/views/create_group_dialog_view.js]: Fixed

File: ts/components/conversation/ConversationHeader.tsx [hunk 7]
```diff
@@ -526,7 +526,7 @@ export class ConversationHeader extends React.Component<Props> {
         {blockHandlerMenuItem}
         {/* {changeNicknameMenuItem}
         {clearNicknameMenuItem} */}
-        {archiveConversationMenuItem}
+        {/* {archiveConversationMenuItem} */}
```
[the remark this answers is not in the document]
Comment (vincentbavitz) [reply ts/components/conversation/ConversationHeader.tsx]: Fixed

File: js/conversation_controller.js [hunk 8]
```diff
@@ -165,6 +165,20 @@
...
+
+        await Promise.all(
+          deviceIds.map(deviceId => {
+            const address = new libsignal.SignalProtocolAddress(id, deviceId);
+            const sessionCipher = new libsignal.SessionCipher(
+              textsecure.storage.protocol,
+              address
+            );
+
+            return sessionCipher.deleteAllSessionsForDevice();
```
[the remark this answers is not in the document]
Comment (vincentbavitz) [reply js/conversation_controller.js]: Fixed; unnecessary

[the remark this answers is not in the document]
Comment (vincentbavitz) [reply js/conversation_controller.js]: Fixed; unnecessary

File: js/views/create_group_dialog_view.js (hunk 5 above)
Comment (neuroscr) [reply js/views/create_group_dialog_view.js:198]: talked in VC, it wouldn't properly await without this structure

**Pull request merged**